teps/0125-configmaps-and-secrets-as-param-value-source.md

```diff
@@ -0,0 +1,189 @@
+- `MAVEN_IMAGE` : Either of the above mentioned solution  doesn't work as the value isn't a param but as an env variable inside the container
+- `CACERTFILE` : Secret doesn't work out of the box as to access env variable `$(params.CACERTFILE)` needs to be replaced with `$CACERTFILE`
+
+### Goals
+
+To reference `PipelineRun` or `TaskRun` param value from ConfigMap or Secret. 
+
+It has following advantages
+- All values passed either as value in param or referenced from `ConfigMap / Secret` can be consistently accessed in a task or all tasks in a pipeline.
+-  Offloads the source of value in a param to its usage in `Pipeline` and `Task`  giving the option to decide the param value on creation of `PipelineRun` or `TaskRun`.
+- Minimises changing of `Pipeline` or `Task` based on source of the value when executing in multiple clusters / namespace promoting reusability.
+
+### Non-Goals
+
+
+## Requirements
+
+- Users should be able to pass `ConfigMap` or `Secret` reference as part of `TaskRun` or `PipelineRun` params along with value.
+
+## Proposal
+
+To achieve the requirement, lets take inspiration from `Kubernetes API Env section of each Container`
+
+```go
+// EnvVar represents an environment variable present in a Container.
+
+type  EnvVar  struct {
+	Name string  `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Value string  `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+type  EnvVarSource  struct {
+	FieldRef *ObjectFieldSelector `json:"fieldRef,omitempty" protobuf:"bytes,1,opt,name=fieldRef"`
+	ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty" protobuf:"bytes,2,opt,name=resourceFieldRef"`
+	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+	SecretKeyRef *SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+
+```
+
+From context of `PipelineRun` and `TaskRun` params 
+```yaml
+	- FieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference field of the pod.
+	- ResourceFieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference resource of the container.
+	- ConfigMapKeyRef : To reference ConfigMap value
+	- SecretKeyRef : To reference Secret value
+```
+
+### Contract
+
+### API
+
+To support the feature supported outlined above, we also propose a small addition to the API. 
+In `ParamSpec` need to `TaskRef` and `PipelineRef` objects will include a `ValueFrom` field.
+
+```go
+type  Param  struct {
+	Name string  `json:"name"`
+	Value ArrayOrString `json:"value"`
+	// Additional field ValueFrom to fetch value from ConfigMap or Secret
+	ValueFrom *ValueSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
```

Comment:
This **Param** struct is of **PipelineRun** / **TaskRun**
not **Pipeline** / **Task**

Reply:
Yes, in `Task` and `Pipeline` we use [`ParamSpec`](https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/apis/pipeline/v1/param_types.go#L44), in `TaskRun` and `PipelineRun` we use [`Param`](https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/apis/pipeline/v1/param_types.go#L115)

Reply:
Yes, indeed, the TEP is about TaskRun and PipelineRun.

If you look at the [specification of a `PipelineRun`](https://github.com/tektoncd/pipeline/blob/6db52ca0c2de48063f4f8a31fea131d72449676e/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L234), you can it uses a type called [`Param`](https://github.com/tektoncd/pipeline/blob/6db52ca0c2de48063f4f8a31fea131d72449676e/pkg/apis/pipeline/v1/param_types.go#L115), which has a name and a [`ParamValue`](
https://github.com/tektoncd/pipeline/blob/6db52ca0c2de48063f4f8a31fea131d72449676e/pkg/apis/pipeline/v1beta1/param_types.go#L132)

The code in your example specifies `ArrayOrString` for the type of value, but that is not true anymore, `value` is now of type `ParamValue`.

```go
type  Param  struct {
  Name string  `json:"name"`
  Value ArrayOrString `json:"value"`
  // Additional field ValueFrom to fetch value from ConfigMap or Secret
  ValueFrom *ValueSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
}
```

It would be good to define the behaviour for parameters of type array and object